rfcs/0053-reducing-daemon-memory-usage.md

```diff
@@ -0,0 +1,137 @@
+
+Below is a list of all the interactions the daemon has with ledger proofs, zkApps proofs, and zkApps verification keys:
+
+* ledger proofs are stored upon receipt of snark work from the gossip network
+* zkApps proofs and newly deployed zkApps verification keys are stored upon receipt of zkApps transactions from the gossip network
+* ledger proofs, zkApps proofs, and zkApps verification keys are read when a block is produced, in order for them to be included into the staged ledger diff
+* ledger proofs, zkApps proofs, and zkApps verification keys are stored upon receipt of new blocks from the gossip network
+* ledger proofs, zkApps proofs, and zkApps verification keys are read when a block is applied to the ledger
+* ledger proofs, zkApps proofs, and zkApps verification keys are read when serving bootstrap requests to nodes joining the network
+
+In order to write this data to disk, we will use the [Lightning Memory-Mapped Database](http://www.lmdb.tech/doc/) (LMDB for short). LMDB is a lightweight, portable, and performant memory map backed key-value storage database. We will demonstrate that the performance of this database is more than sufficient for our use case in the [impact analysis section](#impact-analysis).
+
+For this optimization, it is not important that the on-disk cache is persisted across daemon runs. Such a feature can be added in the future, and this storage layer can double as a way to better persist data from the snark pool and data references from the persistent frontier. However, for now, we will set up the daemon so that it wipes the existing on-disk cache between every restart, in order to simplify the implementation and avoid having to deal with potentially a corrupted on-disk cache (in the event the daemon or operating system did not shut down properly). This is particularly important given our choice of LMDB does not provide complete guarantees against data corruption out of the box, due to the fact memory maps can lead to partial writes if the kernel panics or is otherwise interrupted before the system is gracefully shutdown.
+
+To prevent disk leaks in the cache, we will use GC finalizers on cache references to count the active references the daemon has to information written to the cache. Since the daemon is always starting with a fresh on-disk cache, this will give an accurate reference count to any data cached on-disk. When a GC finalizer decrements the total reference count of an item stored on the cache to 0, it will delete that item from the cache. With this setup, the on-disk cache can only leak if there is a memory leak within the daemon itself, in which the daemon is leaking references to the cache.
```

Comment:
I like that this approach ties the resource management lifetimes together. This feels like a good idea to me.